ch4/shm: Support topology-aware SHM communication

yfguo opened the pull request:
## Pull Request Description

This PR adds the support of detecting node topology and runtime selection of regular/stream memcpy. The PR has four parts:
1. Fixing existing MPMC queue and adding SSE2 version of stream copy for x86 arch without AVX support.
2. Adding node topology detection. Each rank query topology with hwloc and allgather during SHM_post_init. Each rank calculates the topo distance of all its local ranks.
3. Making SHMEM pool to support multiple free queues with different queue types. The number of cells per proc parameter is splited evenly among all free queues. Cell allocation function takes a queue id to indicate which free queue being used in allcoation. No change needed to cell deallcation.
4. Add topology aware SHM communication. A new boolean env `MPIR_CVAR_CH4_SHM_POSIX_TOPO_ENABLE` is added to control this feature. It is disabled by default.

The benefit of using receiver side free queue and non-temporal (stream) store is speed up the inter-NUMA communication (or inter-L3-cache). But non-temporal store is bad for ranks that share the same L3 cache. By making the POSIX topology aware, we can choose non-temporal store inter-NUMA and inter-cache communications.

### Performance
Setup: ALCF Sunspot, Intel Sapphire Rapids x 2. Intel ICX compiler
Configuration: ` --with-device=ch4:ofi --disable-fortran --disable-romio CC=clang CXX=clang++ --enable-fast=all,O3,avx
`
#### `osu_latency`
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{background:#FFC000;
	mso-pattern:black none;}
.xl66
	{background:#E2EFDA;
	mso-pattern:black none;}
.xl67
	{background:#FFF2CC;
	mso-pattern:black none;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


  | TOPO_ENABLE=0 | TOPO_ENABLE=0 | TOPO_ENABLE=1 | TOPO_ENABLE=1
-- | -- | -- | -- | --
Msg Size | intraNUMA | interNUMA | intraNUMA | interNUMA
1 | 0.58 | 1.07 | 0.59 | 1.05
2 | 0.58 | 1.06 | 0.58 | 1.04
4 | 0.58 | 1.05 | 0.59 | 1.04
8 | 0.58 | 1.06 | 0.58 | 1.06
16 | 0.59 | 1.09 | 0.58 | 1.06
32 | 0.59 | 1.06 | 0.59 | 1.09
64 | 0.64 | 1.15 | 0.64 | 1.18
128 | 0.63 | 1.23 | 0.68 | 1.34
256 | 0.67 | 1.28 | 0.72 | 1.44
512 | 0.74 | 1.43 | 0.74 | 1.72
1024 | 0.85 | 1.57 | 0.84 | 1.79
2048 | 1.05 | 1.92 | 1.08 | 1.9
4096 | 1.35 | 2.45 | 1.48 | 2.29
8192 | 2.04 | 3.48 | 2.23 | 2.86
16384 | 4.67 | 8.51 | 5.05 | 6.46
32768 | 6.56 | 12.54 | 6.7 | 8.64
65536 | 10.4 | 21.29 | 10.62 | 13.5
131072 | 18.03 | 37.82 | 18 | 23.24
262144 | 30.93 | 67.66 | 30.86 | 42.08
524288 | 50.96 | 121.63 | 50.95 | 80.75
1048576 | 103.21 | 238.92 | 104.53 | 166.44
2097152 | 220.66 | 500.53 | 224.46 | 348.91
4194304 | 496.01 | 1028.7 | 503 | 698.85
8388608 | 1052.28 | 2085.34 | 1071.69 | 1401.71
16777216 | 2158.29 | 4196.88 | 2186.18 | 2757.43
33554432 | 5449.36 | 8417.38 | 5632.25 | 5425.01
67108864 | 12167.23 | 17285.89 | 12387.28 | 14177.96
134217728 | 25199.15 | 35931.72 | 25594.35 | 31576.1



</body>

</html>

#### `osu_bw`
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{background:#FFC000;
	mso-pattern:black none;}
.xl66
	{background:#E2EFDA;
	mso-pattern:black none;}
.xl67
	{background:#FFF2CC;
	mso-pattern:black none;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


  | TOPO_ENABLE=0 | TOPO_ENABLE=0 | TOPO_ENABLE=1 | TOPO_ENABLE=1
-- | -- | -- | -- | --
Msg Size | intraNUMA | interNUMA | intraNUMA | interNUMA
1 | 4.78 | 1.81 | 4.45 | 1.99
2 | 10.31 | 3.79 | 8.73 | 3.95
4 | 20.22 | 7.71 | 17.84 | 7.75
8 | 39.87 | 15.22 | 35.68 | 16.81
16 | 80.7 | 29.2 | 71.08 | 31.76
32 | 181.01 | 59.5 | 142.21 | 70.94
64 | 320.14 | 118.57 | 240.56 | 132.9
128 | 675.7 | 252.36 | 471.39 | 307.58
256 | 1059.65 | 502.16 | 897.68 | 604.37
512 | 1944.73 | 923.67 | 1692.81 | 476.3
1024 | 3404.64 | 1148.46 | 2966.03 | 861.42
2048 | 3893.98 | 1707.74 | 4403.79 | 1689.73
4096 | 4872.1 | 2649.02 | 5273.76 | 3184.43
8192 | 6435.87 | 3708.21 | 6759.22 | 5681.12
16384 | 5895.42 | 2966.55 | 6377.65 | 5384.48
32768 | 6350.58 | 3092.99 | 6573 | 6558.38
65536 | 6529.19 | 3265.93 | 6856.04 | 7306.86
131072 | 6778.2 | 3418.82 | 7052.65 | 7899.33
262144 | 6919.02 | 3527.45 | 7201.67 | 8034.9
524288 | 6997.38 | 3591.29 | 7393.26 | 8237.02
1048576 | 8090.02 | 4053.7 | 8279.48 | 9055.35
2097152 | 7296.57 | 3940.11 | 7022.39 | 10202.08
4194304 | 7320.59 | 3916.27 | 5989.3 | 8275.19
8388608 | 7324.49 | 3967.97 | 6111 | 8852
16777216 | 7306.44 | 3973.63 | 6509.18 | 9186.05
33554432 | 7321.26 | 3973.26 | 7239.55 | 9348.05
67108864 | 6021.43 | 3968.32 | 6206.13 | 9389.08
134217728 | 5393.47 | 3855.42 | 5611.83 | 5941.02



</body>

</html>



## Author Checklist
* [ ] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [ ] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [ ] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

hzhou commented on configure.ac:
```diff
@@ -891,6 +891,7 @@ for option in $enable_fast ; do
     case "$option" in
         O*)
         enable_fast_opts=$option
+        enable_fast_sse_instr=yes
```
It doesn't make sense to set the default here. Move it outside.

yfguo replied:
WIll fix.

hzhou commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.c:
```diff
@@ -31,6 +31,16 @@
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        Control if sender side or receiver side free cell queue is used.
+
```
Don't move CVAR in this commit since you are going to delete it the next commit anyway.

yfguo replied:
WIll fix.

hzhou commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.c:
```diff
@@ -56,12 +46,15 @@ static int init_transport(int vci_src, int vci_dst)
-    if (MPIR_CVAR_CH4_SHM_POSIX_IQUEUE_FREE_QUEUE_SENDER_SIDE) {
-        int queue_type = MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC;
-        mpi_errno = MPIDU_genq_shmem_pool_create(transport->size_of_cell, transport->num_cells,
+    if (MPIR_CVAR_CH4_SHM_POSIX_TOPO_ENABLE) {
```
Wait, so the else branch -- `MPIR_CVAR_CH4_SHM_POSIX_TOPO_ENABLE=0` -- will use `__MPMC` free queue?

yfguo replied:
No. It is a typo. Will fix.

hzhou reviewed (commented):
Looks good. Only a few minor comments. Please also add commit messages explaining the changes in addition to commit titles.

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

yfguo commented:
Looking at the `manyrma2` failure.

hzhou commented:
> Looking at the `manyrma2` failure.

I think it is just a time-out due to active message path being too slow. It is not related to your PR. 

Could you update the performance measurement since the last one uses MPMC free queue in the `TOPO_ENABLE=0`?

Also, please add commit messages in addition to a single line of title. The commit message should explain the changes, such as why and what.

yfguo commented:
I have updated the commit messages and rebased on the latest `main`. The results in the PR is also updated.

yfguo commented:
test:mpich/ch4/ofi

hzhou commented:
> I have updated the commit messages and rebased on the latest `main`. The results in the PR is also updated.

Thanks! The difference between enable and disable `TOPO_ENABLE` on intranuma are attributable to noise, right?

hzhou reviewed (approved):
Approved. Please fix the spelling error, check the PR author checklist, and clear the tests before merging.

Thank you for making the changes! Good PR!

raffenet commented on src/mpid/ch4/shm/posix/posix_init.c:
```diff
@@ -148,6 +161,61 @@ static int init_vci(int vci)
+    hwloc_cpuset_t cpuset;
+    MPIDI_POSIX_topo_info_t *topo = &MPIDI_POSIX_global.topo;
+
+    hwloc_topology_init(&topology);
+    hwloc_topology_load(topology);
```
can we use the `MPIR_hwtopo` API instead? another hwloc init could cause startup performance issues.

yfguo replied:
Sure. I can do that.

yfguo commented:
> > I have updated the commit messages and rebased on the latest `main`. The results in the PR is also updated.
> 
> Thanks! The difference between enable and disable `TOPO_ENABLE` on intranuma are attributable to noise, right?

Yes. Those are noises.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
@raffenet I have redo the topo detection code with the MPIR_hwtopo APIs.

raffenet commented on src/mpid/ch4/shm/posix/posix_init.c:
```diff
@@ -148,6 +161,18 @@ static int init_vci(int vci)
+    MPIDI_POSIX_topo_info_t *topo = &MPIDI_POSIX_global.topo;
+    MPIR_hwtopo_gid_t core_id = MPIR_hwtopo_get_obj_by_type(MPIR_HWTOPO_TYPE__CORE);
+    MPIR_hwtopo_gid_t l3_cache_id = MPIR_hwtopo_get_obj_by_type(MPIR_HWTOPO_TYPE__L3CACHE);
+    MPIR_hwtopo_gid_t numa_id = MPIR_hwtopo_get_obj_by_type(MPIR_HWTOPO_TYPE__PACKAGE);
```
It looks like hwloc still defines `HWLOC_OBJ_NUMANODE`, our interface just doesn't provide an equivalent for it. Can you add support for it and use it here instead of package to avoid any possible mismatch? https://www.open-mpi.org/projects/hwloc/doc/v2.11.0/a00140.php#ggacd37bb612667dc437d66bfb175a8dc55a9d917a3e5497950c6d8948b8e183db5a

raffenet replied:
I guess in our abstraction `MPIR_HWTOPO_TYPE__DDR` is equivalent to `HWLOC_OBJ_NUMANODE`.

raffenet replied:
```suggestion
    MPIR_hwtopo_gid_t numa_id = MPIR_hwtopo_get_obj_by_type(MPIR_HWTOPO_TYPE__DDR);
```

hzhou commented on src/mpid/ch4/shm/posix/posix_init.c:
```diff
@@ -42,11 +52,14 @@
 #include "utarray.h"
 #include <strings.h>    /* for strncasecmp */
 
+#include <hwloc.h>
```
use `mpir_hwtopo.h` instead.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
Tests were clean. Rebased on latest `main`.
